hooks/basic/collector.py

```diff
@@ -214,12 +216,84 @@ def on_context_changed(self, settings, item):
+                work_tmpl = self.sgtk.template_from_path(path)
+
+                if not work_tmpl:
+                    # this template was not found in the template config!
+                    raise TankError("The template '%s' does not exist!" % work_path_template)
+
+                # update the field with correct value so that we can use it everytime for this item
+                work_path_template = work_tmpl.name
+
+        # Else see if the path matches an existing template
+        else:
+            # let's try to check if this path fits into any known template
+            work_tmpl = self.sgtk.template_from_path(path)
+            if not work_tmpl:
+                # this path doesn't map to any known templates!
+                raise TankError("Cannot find a matching template for path: %s" % path)
+
+            # update the field with correct value so that we can use it everytime for this item
+            work_path_template = work_tmpl.name
+
         if work_path_template:
             item.properties["work_path_template"] = work_path_template
 
         # Set the item's fields property
         item.properties["fields"] = self._resolve_item_fields(item)
 
+    def _build_context(self, item):
+        """Updates the context of the item from the work_path_template/template, if needed.
+
+        :param item: item from publisher.
+        """
+
+        publisher = self.parent
+
+        # this should be defined and correct by now!
+        # Since we resolve this field too, while context change of the item.
+        work_path_template = item.properties.get("work_path_template")
+        path = item.properties.get("path")
+
+        work_tmpl = publisher.get_template_by_name(work_path_template)
+
+        # evaluate the context of the item too at the time of addition.
+        # user can change this later if they don't like it.
+        # this is being done here, since the tasks are created after the process_file hook runs.
+        # We anyways can't resolve a TASK context using context_from_path?!
+        # So we can check for that? In case we are launching the publisher with correct context.
+        if not item.context.task:
+            new_context = self.tank.context_from_path(path)
+            entities = work_tmpl.get_entities(path)
+            entity_dict = {entity['type'].lower(): entity for entity in entities}
+            valid_entity_dict = None
+
+            if 'asset' in entity_dict:
+                valid_entity_dict = context._get_valid_entity_dict(self.tank, entity_dict['asset'])
+            elif 'shot' in entity_dict:
+                valid_entity_dict = context._get_valid_entity_dict(self.tank, entity_dict['shot'])
+            elif 'sequence' in entity_dict:
+                valid_entity_dict = context._get_valid_entity_dict(self.tank, entity_dict['sequence'])
+            elif 'project' in entity_dict:
+                valid_entity_dict = context._get_valid_entity_dict(self.tank, entity_dict['project'])
```

Comment:
This is overkill and expensive. I've created a new api method instead to make this easier/faster. Try `tk.context_from_entities(entities)` in `sgtk_core-2.11.0`

Reply:
This is still not working for me, since the context that I get from the method doesn't have a **Step** entity.

Reply:
Opened a separate pull request for this, where I'm adding 'step' to the context if it's not already there.